### PR TITLE
feat: enchance pedersen instance def

### LINKED ISSUE #217: feat: Pedersen Instance Definition Enhancement
## Description
The current Pedersen Instance Definition in our Cairo Zig codebase requires enhancements: [Pedersen Instance Definition](https://github.com/keep-starknet-strange/cairo-zig/blob/main/src/vm/types/pedersen_instance_def.zig).

## Objectives
- Implement improvements within the Pedersen Instance Definition.
- Enhance test coverage for thorough validation and accuracy.

## References
- [Rust Implementation Reference](https://github.com/lambdaclass/cairo-vm/blob/8887e60f7ea5459188cb6624acc908ad29e75f60/vm/src/types/instance_definitions/pedersen_instance_def.rs)
- [Python Implementation Reference](https://github.com/starkware-libs/cairo-lang/blob/51c287ac441ea0295775dce743470f87e395d4b1/src/starkware/cairo/lang/builtins/hash/instance_def.py)

### PR DESCRIPTION
**Pull request by dhruvkelawala**
closes #217

### CODE REVIEW

File: src/vm/types/pedersen_instance_def.zig
Comment (tcoratger) [inline src/vm/types/pedersen_instance_def.zig:68]: This one is declared as public constant so that we don't need a dedicated function for that in my opinion.

File: src/vm/types/pedersen_instance_def.zig
Comment (tcoratger) [inline src/vm/types/pedersen_instance_def.zig:45]: Can you add a comma after the last item of the list for better formatting?

File: src/vm/types/pedersen_instance_def.zig
Comment (dhruvkelawala) [reply src/vm/types/pedersen_instance_def.zig:35]: Yes, thanks for letting me know. This seems like a very good change. Using BigInt seems quite a hassle

**Comment by dhruvkelawala** [comment]
@tcoratger Should I wait for your PR to be merged for refactoring to u256?

**Comment by tcoratger** [comment]
> @tcoratger Should I wait for your PR to be merged for refactoring to u256?

@dhruvkelawala No you can do that before let us just wait for @lana-shanghai approval in order to confirm that this is the good way to go.

**Comment by edisontim** [comment]
Can we get rid of all the variables except for ratio? That's the only one used in practice in the cairo-vm Rust codebase

**Comment by dhruvkelawala** [comment]
> Can we get rid of all the variables except for ratio? That's the only one used in practice in the cairo-vm Rust codebase

Wdym? We don't need hash_limit and repetitions?

**Comment by edisontim** [comment]
Yeah I believe none of the variables except for ratio are used

File: src/vm/types/pedersen_instance_def.zig
Comment (lana-shanghai) [reply src/vm/types/pedersen_instance_def.zig:35]: Hey @dhruvkelawala yes since we use Felt252 then u256 should be enough to handle our field elements. Feel free to simplify the PR!

**Comment by lana-shanghai** [comment]
> Yeah I believe none of the variables except for ratio are used

Let's keep it as is for now and then later take a bigger overview if we need them. I would prefer to have the same structure of each builtin even if we aren't using one of the fields to do general switch patterns if needed etc.

File: src/vm/types/pedersen_instance_def.zig
Comment (lana-shanghai) [reply src/vm/types/pedersen_instance_def.zig:83]: Yes this might be a bit redundant

File: src/vm/types/pedersen_instance_def.zig
Comment (lana-shanghai) [inline src/vm/types/pedersen_instance_def.zig:34]: Please sync with @jobez regarding naming conventions and initializing default structs.

**Comment by edisontim** [comment]
> > Yeah I believe none of the variables except for ratio are used
> 
> Let's keep it as is for now and then later take a bigger overview if we need them. I would prefer to have the same structure of each builtin even if we aren't using one of the fields to do general switch patterns if needed etc.

That's fine for me but currently the structs in the cairo-vm-rust as well as the keccak in zig-vm aren't using the same members: 
- Pedersen: https://github.com/lambdaclass/cairo-vm/blob/8a2ef24c1ca258f1d1d230cbbd1478cd1d4cf9db/vm/src/types/instance_definitions/pedersen_instance_def.rs#L20
- Keccak: https://github.com/lambdaclass/cairo-vm/blob/8a2ef24c1ca258f1d1d230cbbd1478cd1d4cf9db/vm/src/types/instance_definitions/keccak_instance_def.rs#L6

File: src/vm/types/pedersen_instance_def.zig
Comment (jobez) [reply src/vm/types/pedersen_instance_def.zig:34]: hey! lets go with default -> initDefault

see https://github.com/ziglang/zig/blob/master/lib/std/enums.zig#L737 as prior art

File: src/vm/types/pedersen_instance_def.zig
Comment (jobez) [inline src/vm/types/pedersen_instance_def.zig:34]: what do you think about using zig's ability to set default field values here?

I am thinking:

for values that can be known at comptime, we can set them like so https://ziglang.org/documentation/master/#Default-Field-Values

and then just have `initDefault` return `.{}`

Comment (tcoratger) [reply src/vm/types/pedersen_instance_def.zig:34]: Agreed

Comment (jobez) [reply src/vm/types/pedersen_instance_def.zig:34]: ok, so the style is:

initDefault for structs if and only if theres an allocator involved, otherwise we use zig's default field values

Comment (lana-shanghai) [reply src/vm/types/pedersen_instance_def.zig:34]: so we are waiting for this minor fix @jobez and then merging right? 
@dhruvkelawala I think this is the only little thing left

File: src/vm/types/pedersen_instance_def.zig
Comment (lana-shanghai) [reply src/vm/types/pedersen_instance_def.zig:46]: ping @dhruvkelawala

Comment (dhruvkelawala) [reply src/vm/types/pedersen_instance_def.zig:46]: Should I also remove `rangeCheckPerBuiltin`?

**Review by lana-shanghai — APPROVED** [review]
lgtm

**Pull request merged**
